### DISCUSSION TITLE
Handle Multiple rank arrays datatypes

### DISCUSSION THREAD

**Discussion by said-hadjout** [pyccel/pyccel#1737] [Ideas]
In the current syntax, the user can only specify an array of a specific rank.
It would be nice if we could handle an array of multiple ranks, something like `int[1:4]` that translates to `int[:]|int[:,:]|int[:,:,:]|int[:,:,:,:]|`.
This will make it possible for us to define something like
```python
@template('T', types=['int8', 'int16', 'int32', 'int64', 'float32', 'float64', 'complex64', 'comples128'])
def sum(a:'T[1:15]'):
    return a.sum()
```
This function can take any array of numerical types with ranks going from 1 to 15 and calculate its sum

**Comment by said-hadjout** [comment]
We can also use something like `int[2:10:2]`  to handle even rank arrays by specifying the `step`

**Comment by yguclu** [comment]
I would like to mention the forgotten `@elemental` decorator...

**Reply by said-hadjout** [reply]
how can we compute the sum of an array using elemental?

**Reply by yguclu** [reply]
We can not! 😅 

But if you are interested in a user-defined reduction operation similar to `sum`, how can you write it down explicitly in Python? The number of loops will change depending on the number of dimensions, so we may need the concept of a generic iterator in order to avoid explicit loops.

**Reply by said-hadjout** [reply]
we can use `numpy.ndindex` and have something like
```python
for index in np.ndindex(a.shape):
    print(index)
(0, 0, 0)
(0, 1, 0)
(1, 0, 0)
(1, 1, 0)
(2, 0, 0)
(2, 1, 0)
....
```

**Reply by said-hadjout** [reply]
we also need a way to represent multi-dimensional slices

**Reply by EmilyBourne** [reply]
> we also need a way to represent multi-dimensional slices

In what language? If anything other than Python, then, why?

**Reply by saidctb** [reply]
I want to be able to do something like this, but I don't know how
```python
def f(a:'int[:,:]|int[:,:,:]'):
    s1 = a[...,0].sum()
    s2 = a[1:,1:,1:].sum() if rank(a)==3 elif rank(a) ==2 a[1:,1:].sum()
    return s1,s2
```

**Reply by saidctb** [reply]
maybe we can handle something like
```python
I = tuple(slice(1,None) for _ in range(len(a.shape)))
print(a[I].sum())
```

**Reply by EmilyBourne** [reply]
I guess `s1` should be possible after your PR ?
`s2` doesn't work in pyccel currently. Do you have a use case in mind? This seems like a corner case to me. If NumPy don't have syntax for it then I tend to think it's not a very common usage. Unless you have an example of a place where this kind of slicing is commonly used?

**Reply by EmilyBourne** [reply]
It wouldn't be hard to support though. But I guess you would need something equivalent to a `if constexpr` in C++ to indicate that you want the expression to be simplified

**Reply by EmilyBourne** [reply]
> maybe we can handle something like
> ```python
> I = tuple(slice(1,None) for _ in range(len(a.shape)))
> print(a[I].sum())
> ```

What happens with this currently? It fails because we don't store slice objects?

**Reply by EmilyBourne** [reply]
I think it would be simple to support this. But we need an issue for saving a Slice in a Variable

**Reply by saidctb** [reply]
> I guess `s1` should be possible after your PR ? `s2` doesn't work in pyccel currently. Do you have a use case in mind? This seems like a corner case to me. If NumPy don't have syntax for it then I tend to think it's not a very common usage. Unless you have an example of a place where this kind of slicing is commonly used?

s1 is not handled in my PR, I just did the naive case of `a[...]`
as for s2 we use that a lot in Psydac in [here](https://github.com/pyccel/psydac/blob/devel/psydac/linalg/stencil.py#L2266) for example and this [file](https://github.com/pyccel/psydac/blob/devel/psydac/linalg/kernels/matvec_kernels.py) is just the expansion of the `_dot` method

**Reply by saidctb** [reply]
I don't know if we support the syntax `tuple( i for i in range(10))`, as for the `slice` , we need to have an object to store `(start, end, step)` , and we need to handle the None value in`slice(None,None)`

**Reply by EmilyBourne** [reply]
I've just checked. We don't, but we support `[i for i in range(10)]` so it is fairly trivial to add we just need an issue

**Comment by said-hadjout** [comment]
we can use this syntax `int[1&...&4]` to specify the range of ranks
